matchtree.go

```diff
@@ -213,29 +210,29 @@ func (t *symbolSubstrMatchTree) prepare(doc uint32) {
 		fileStart = t.fileEndRunes[doc-1]
 	}
 
-	for len(t.sections) > 0 && t.sections[0].Start < fileStart {
-		t.sections = t.sections[1:]
-		t.secID++
+	var sections []DocumentSection
+	if len(t.sections) > 0 {
+		sections = t.sections[t.fileEndSymbol[doc]:t.fileEndSymbol[doc+1]]
 	}
+	secIdx := 0
 
-	// TODO we can use fileEndSymbol to skip
 	trimmed := t.current[:0]
-	for len(t.sections) > 0 && len(t.current) > 0 {
+	for len(sections) > secIdx && len(t.current) > 0 {
```

Comment:
IMPORTANT: I didn't realise this when I first reviewed, but you are doing far too much work in prepare. Prepare should pretty much be O(1). You should only do this work in matches.

IE if you do a search like `f:foo sym:bar` Prepare will run for every document. However, matches won't run with a high enough cost for symbol since f:foo matchtree will rule out most documents.

So essentially this whole function should be moved into matches, and prepare should only reset fields in the struct.

Reply:
Gotcha, will change. I initially placed it here to resemble upstream zoekt (which does this work in prepare`).

Reply:
I might be misunderstanding Zoekt, but I don't think prepare runs on every document. It runs only on documents returned by nextDoc (so a document that has foo in its name and bar in its content).

```go
// An expression tree coupled with matches. The matchtree has two
// functions:
//
// * it implements boolean combinations (and, or, not)
//
// * it implements shortcuts, where we skip documents (for example: if
// there are no trigram matches, we can be sure there are no substring
// matches). The matchtree iterates over the documents as they are
// ordered in the shard.
//
// The general process for a given (shard, query) is
//
// - construct matchTree for the query
//
// - find all different leaf matchTrees (substring, regexp, etc.)
//
// in a loop:
//
//   - find next doc to process using nextDoc
//
//   - evaluate atoms (leaf expressions that match text)
//
//   - evaluate the tree using matches(), storing the result in map.
//
//   - if the complete tree returns (matches() == true) for the document,
//     collect all text matches by looking at leaf matchTrees
//
type matchTree interface {
	docIterator

	// returns whether this matches, and if we are sure.
	matches(cp *contentProvider, cost int, known map[matchTree]bool) (match bool, sure bool)
}
```

Reply:
Yes, `prepare` is only called on `nextDoc` which means it should generally only be prepared on documents which actually have matches. But just because a matchtree returns a doc from nextDoc doesn't imply it actually matches. As a stark example try `f:go$` then nextDoc can't rely on trigram indexes.

But yeah given our trigram indexes work so well this isn't as important as I thought! Sorry for delaying this PR. Rather ship this idea in a follow up (and possibly upstream?)

Reply:
Thanks for explaining! I'll follow up on this in a separate PR.